rinja_derive/src/generator.rs

```diff
@@ -1962,28 +1919,59 @@ struct Buffer {
     // The buffer to generate the code into
     buf: String,
     discard: bool,
+    last_was_write_str: bool,
 }
 
 impl Buffer {
     fn new() -> Self {
         Self {
             buf: String::new(),
             discard: false,
+            last_was_write_str: false,
         }
     }
 
+    fn is_discard(&self) -> bool {
+        self.discard
+    }
+
+    fn set_discard(&mut self, discard: bool) {
+        self.discard = discard;
+        self.last_was_write_str = false;
+    }
+
     fn writeln(&mut self, src: impl BufferFmt) {
         if !self.discard {
             src.append_to(&mut self.buf);
             self.buf.push('\n');
+            self.last_was_write_str = false;
         }
     }
 
     fn write(&mut self, src: impl BufferFmt) {
         if !self.discard {
             src.append_to(&mut self.buf);
+            self.last_was_write_str = false;
         }
     }
+
+    fn write_writer(&mut self, s: &str) -> usize {
+        if self.discard {
+            // nothing to do
+        } else if !self.last_was_write_str {
+            writeln!(self.buf, "writer.write_str({s:#?})?;").unwrap();
+            self.last_was_write_str = true;
+        } else {
+            // strip trailing `")?\n`, leaving an unterminated string
+            let len = self.buf.strip_suffix("\")?;\n").unwrap().len();
+            self.buf.truncate(len);
+            // append the new string, adding a stray `"` in the mid of the string
+            writeln!(self.buf, "{s:#?})?;").unwrap();
+            // left shift new string by one to overwrite the stray `"`
+            self.buf.replace_range(len..=len, "");
```

Comment:
The longer the string, the worse it gets. I think for this, it'd be better to keep a `String` in `self` and simply `clear` every time, then render into and then push the new string without the first `"` into `buf`. Like that, no need for the shift and it also greatly reduces the number of allocations.

Reply:
Fixed in <https://github.com/rinja-rs/rinja/pull/86/commits/9ba70cbce8ba0ab8db173f52f724becef21c4411>. It has a healthy speed-up for other cases, too. The normal <str as fmt::Debug> is optimized for human readability, but we don't need it to be. Simply dump everything as is, except `\" \\ \r`.